poc/mastic.py

```diff
@@ -139,67 +131,57 @@ def shard_with_joint_rand(
+        (vidpf_rand, rand) = front(self.vidpf.RAND_SIZE, rand)
+        (prove_rand_seed, rand) = front(self.xof.SEED_SIZE, rand)
+        (leader_seed, rand) = front(self.xof.SEED_SIZE, rand)
+        (helper_seed, rand) = front(self.xof.SEED_SIZE, rand)
+        assert len(rand) == 0  # REMOVE ME
 
         (alpha, meas) = measurement
         beta = self.flp.encode(meas)
 
         # Generate VIDPF keys.
-        (vidpf_public_share, vidpf_keys) = \
-            self.vidpf.gen(alpha, beta, nonce, vidpf_gen_rand)
+        (correction_words, keys) = \
+            self.vidpf.gen(alpha, beta, nonce, vidpf_rand)
 
         # Generate FLP joint randomness.
-        joint_rand_parts = []
-        joint_rand_parts.append(self.joint_rand_part(
-            0, flp_leader_seed, vidpf_keys[0], vidpf_public_share, nonce))
-        joint_rand_parts.append(self.joint_rand_part(
-            1, flp_helper_seed, vidpf_keys[1], vidpf_public_share, nonce))
+        joint_rand_parts = [
+            self.joint_rand_part(
+                0, leader_seed, keys[0], correction_words, nonce),
+            self.joint_rand_part(
+                1, helper_seed, keys[1], correction_words, nonce),
+        ]
         joint_rand = self.joint_rand(
             self.joint_rand_seed(joint_rand_parts))
-        flp_public_share = joint_rand_parts
-
-        # Generate FLP proof shares.
-        flp_prove_rand = self.prove_rand(flp_prove_rand_seed)
-        flp_proof = self.flp.prove(beta, flp_prove_rand, joint_rand)
-        flp_leader_proof_share = vec_sub(
-            flp_proof,
-            self.helper_proof_share(flp_helper_seed),
-        )
 
-        public_share = (vidpf_public_share, flp_public_share)
+        # Generate FLP and split it into shares.
+        prove_rand = self.prove_rand(prove_rand_seed)
+        proof = self.flp.prove(beta, prove_rand, joint_rand)
+        helper_proof_share = self.helper_proof_share(helper_seed)
+        leader_proof_share = vec_sub(proof, helper_proof_share)
+
+        public_share = (correction_words, joint_rand_parts)
         input_shares = [
-            (vidpf_keys[0], flp_leader_proof_share, flp_leader_seed),
-            (vidpf_keys[1], None, cast(Optional[bytes], flp_helper_seed)),
+            (keys[0], leader_proof_share, leader_seed),
+            (keys[1], None, cast(Optional[bytes], helper_seed)),
         ]
         return (public_share, input_shares)
 
     def is_valid(self,
                  agg_param: MasticAggParam,
                  previous_agg_params: list[MasticAggParam],
                  ) -> bool:
-        (level, prefixes, do_range_check) = agg_param
-
         # Check that the range check is done exactly once.
-        first_level_range_check = \
-            (do_range_check and len(previous_agg_params) == 0) or \
-            (not do_range_check and
+        range_checked = \
+            (agg_param[2] and len(previous_agg_params) == 0) or \
```

Comment:
`agg_param[2]` makes me go back to see what is `agg_param[0]` and `agg_param[1]`.
Could we keep `(_level, _prefixes, do_range_check) = agg_param` ?

Reply:
Done!